rfd/0073-discover.md

```diff
@@ -0,0 +1,337 @@
+
+### 3. Heartbeat
+
+`teleport discover` will periodically heartbeat discovered resources to the
+cluster's Auth server with some TTL.
+
+The discovery heartbeat will use similar mechanism as used by other agents and
+each discovered resource will be converted to a Teleport resource. For example,
+each EC2 instance will be heartbeat as a `Server` resource, each RDS database
+as a `Database` resource, etc. Each resource will include cloud metadata needed
+to later identify the cloud resource it corresponds to.
+
+To prevent mixing up the resources that are already part of the cluster with
+the resources from the discover flow, the discover heartbeat will keep them
+under a different path in the Auth backend.
+
+Each `teleport discover` run will have a unique ID which will allow to retrieve
+resources discovered during this particular run. This will allow the UI (or the
+CLI in the CLI-only mode) to display the discovered resources to the user.
+
+The ID will be auto-generated as a UUIDv4 or can be optionally specified on the
+command line:
+
+```
+$ teleport discover aws --discover-id=abc123
+```
+
+Every discovered resource will be associated with this ID by being stored under
+a particular path in the backend during the heartbeat, e.g.
+
+```
+/discover/abc123/discovered/server/node-1
+/discover/abd123/discovered/db/mysql
+```
+
+A set of backend APIs will need to be implemented to enable this flow, as well
+as a web API handler for the UI to retrieve discovered resources:
+
+```
+GET /webapi/sites/:site/discover/:id/servers
+GET /webapi/sites/:site/discover/:id/databases
+...
+// returns a list of JSON-dumped resources
+```
+
+### 4. Confirmation
+
+After launching the discover and heartbeat service, the discover command will
+block waiting for the user to inspect the resources in the UI and select those
+they want to connect.
+
+All resources selected by the user will be saved under a different path in the
+backend for the same discover run:
+
+```
+/discover/abc123/connect/server/node-1
+/discover/abd123/connect/db/mysql
+```
+
+The web UI will call a new API handler to post resources selected by the user:
+
+```
+POST /webapi/sites/:site/discover/:id/servers
+POST /webapi/sites/:site/discover/:id/databases
+...
+```
+
+The `teleport discover` command will watch the `/discover/<id>/connect` key and
+bring the resources that appear there into the cluster.
+
+### 5. Import
+
+For EC2 nodes `teleport discover` will:
+
+- Use SSM to install a Teleport SSH agent on each EC2 instance using the same
+  approach as described in RFD 57.
+
+For RDS and other AWS databases `teleport discover` will:
+
+- Register discovered databases in Teleport using dynamic resource registration.
+
+Doing so will require appropriate IAM permissions to run SSM commands (for SSH)
+and setup IAM policies (for RDS).
+
+### 6. Shutdown
+
+After importing all resources, `teleport discover` will print out a summary of
```

Comment:
@r0mant you mentioned that you can let users to keep the service running. Can teleport discover generate systemd file, apt-get? Basically can we show users how to keep this stick without going to the docs? Something like that:

```
$ teleport discover isntall

[1] Generated `/etc/teleport.yaml`
[2] Generated `/etc/systemd.d/teleport.service

Run `sudo systemctl teleport start` to keep Teleport agent running. For HA or K8s setup follow this guide:

....

```